### PR TITLE
Issue/5999 compose guidelines

### LINKED ISSUE #5999: Compose guidelines for WCAndroid
There has been some discussion around adding Compose to the project. We've done a proof of the concept while working on the More Menu tab. But it's time to define some guidelines and best practices on how we should with Compose inside WooCommerce. Following topics should be covered in the guidelines: 

- Code styles
- Package structuring of compose files
- State hoisting
- Navigation
- Writing stateless reusable composable functions with single source of truth
- Accesibility

### PR DESCRIPTION
**Pull request by JorgeMucientes**
Closes: #5999 

### Description
The idea of this PR is to open discussion over best practices and style guides on using Compose in WCAndroid. This PR contains the guidelines along with changes to the existing Compose code to comply with those guidelines.  

You are welcome to change or suggest anything you want from this PR. 

You can see the guidelines in a better-looking format [here](https://github.com/woocommerce/woocommerce-android/blob/trunk/docs/compose.md)

### CODE REVIEW

File: docs/compose.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+* Compose Library Development
+* Compose App development
+
+We will apply the **App development** guidelines which are more flexible unless we are developing a Compose specific library, in which case Library development level should be applied. 
+
+Many of the rules from official guidelines are already integrated with Lint, so it will be easier to comply with them. 
+
+Any exception to those code style guidelines should be described here: 
+
+* Guidelines suggest to use `PascalCase` for constant values, Enums, etc. We are currently using `All caps snake_case` for this. There is no major reason for using `PascalCase` in Compose code and have different styles across the app. So, we will keep using `All caps snake_case` for Compose code.
```
Comment (hichamboushaba) [inline docs/compose.md]: Question: do we have this in the conventions somewhere? I didn't look in the docs, just wondering, because from my experience with the code we seem to be using both conventions throughout the code.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt [hunk 2]
```diff
@@ -139,20 +140,17 @@ private fun MoreMenuMyStoreHeader(
             Text(
                 text = siteName,
                 fontWeight = FontWeight.Bold,
-                fontSize = 20.sp,
-                color = colorResource(id = color.color_on_surface)
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt:143]: Removing the color on those parts breaks the app on dark mode:


One way to solve this is to set `setTextColors` to true in the `createMdcTheme` call to read colors from the textAppearances, but this has some bad implications as the docs states: "By default the text colors from any associated `TextAppearance`s from the theme are not read. This is because setting a fixed color in the resulting [TextStyle] breaks the usage of [androidx.compose.material.ContentAlpha] through [androidx.compose.material.LocalContentAlpha].", which means automatic content color resolving won't work.

A better way IMO, and which would match the behavior that happens on pure Compose apps is to have a guideline to use a `Surface` as a parent for all of our Screens, this is because `Surface` can automatically set the content color based on its background (https://developer.android.com/jetpack/compose/themes/material#content-color), just to clarify what I mean, it's using something like this in the fragment:
```kotlin
setContent {
                WooTheme {
                    Surface(color = MaterialTheme.colors.background) {
                        MoreMenu(viewModel)
                    }
                }
            }
```
If you test this you'll notice that the colors of texts are set correctly, if we agree on this, we can then remove the rest of color references (like this [one](https://github.com/woocommerce/woocommerce-android/blob/6b8f11bc5161ec158078cb8aaaf43fb37b819979/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt#L152))

**Review by hichamboushaba — COMMENTED** [review]
Nice job @JorgeMucientes, I swiped through the docs, and they are super nice.

But I'm not approving yet, the changes break the screen on dark mode, please check my comment, and let's discuss it there, if we agree on the solution, it should be added to the guidelines too.

And one remark, I think you just missed some spots, but I noticed we have still some usages of `colorResource(id = color....)`, I think we should replace the ones that follows the Material coloring system to their equivalent: for example `colorResource(id = color.color_primary)` should be replaced by `MaterialTheme.colors.primary`

File: docs/compose.md (hunk 1 above)
Comment (JorgeMucientes) [reply docs/compose.md]: hmm, good point @hichamboushaba , maybe I wrongly assumed that we were following `uppercase_snake_case` for such cases. Our docs say " Our code style guidelines are based on the [Android Code Style Guidelines for Contributors](https://developer.android.com/kotlin/style-guide)". Those guidelines recommend using `uppercase_snake_case` for [constants names](https://developer.android.com/kotlin/style-guide#constant_names). For Enum values, it doesn't really say anything, but all the [examples](https://developer.android.com/kotlin/style-guide#enum_classes) they provide for Enum values are using `upper snake case`. For Singletons and Sealed classes, I'd say to keep using `PascalCase` as we've been using it until now. 
Basically, the idea of that exception is to state that we should keep using the code style conventions we were using up until now instead of forcing `PascalCase` for constant and enum values. I'll rephrase it if you agree with this.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt (hunk 2 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt:143]: Nice catch on the broken dark theme @hichamboushaba, and thanks for the input. 

> "By default the text colors from any associated TextAppearances from the theme are not read. This is because setting a fixed color in the resulting [TextStyle] breaks the usage of [androidx.compose.material.ContentAlpha] through [androidx.compose.material.LocalContentAlpha]."

TIL. Ok given that is the case, I think setting up a surface as parent for all the content makes sense. I was not aware about the existence of `surface` and its behavior either. One thing I don't like that much about that approach is having to set the  `Surface` as root every time we want to set the content for a composable view. What do you think about "wrapping" the content with a `surface` inside `WooTheme` ?? Something like this: 

```kotlin 
@Composable
fun WooTheme(
    content: @Composable () -> Unit
) {
    val (colors, typography, shapes) = createMdcTheme(
        context = LocalContext.current,
        layoutDirection = LocalLayoutDirection.current
    )

    BaseWooTheme(
        colors = colors,
        typography = typography,
        shapes = shapes,
        content = { SurfacedContent(content) },
    )
}

@Composable
private fun SurfacedContent(
    content: @Composable () -> Unit
) {
    Surface(color = MaterialTheme.colors.background) {
        content()
    }
}
```
On one hand we are hiding the use of the surface inside the theme so it is not that explicit what is going on. But on the other hand, we avoid any risk of forgetting about setting the surface and we do not have to create an extra guideline for it. Unless I'm missing any potential side effect I lean more towards this last solution. 
Wdyt?

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt:143]: > On one hand we are hiding the use of the surface inside the theme so it is not that explicit what is going on

This is the main issue I have with this solution, and in addition to this, I feel like it's not the responsibility of the `WooTheme` to add a parent composable.
Also, the `Surface` parent is needed only when the screen doesn't use a parent that doesn't use an explicit parent that supports setting a content color, for example, if the screen was using a `Scaffold` (like the About library), the `Surface` composable won't be needed.

WDYT about inverting the roles a bit? I mean keeping `WooTheme` as it was before, but we add a new variant called `WooThemWithBackground` (or something similar) that adds the `Surface` parent, and this way it'll be up to the Fragment to decide which variant works best for their case.

Anyway, this is not a blocker, I think we can revisit whatever decision we make, and it won't be hard to refactor (the number of screens won't be that big), we just need to make sure it's documented, so if you decide to leave as you have it now, just add a comment explaining why the `Surface` parent is used 🙏.

Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt:143]: I agree with the suggested approach. I've updated the code and the guidelines accordingly. Let me know if you miss anything. 

Thanks once again for going over this and providing great input @hichamboushaba

**Review by hichamboushaba — APPROVED** [review]
Looks great, thanks so much @JorgeMucientes for wrangling this

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt (hunk 2 above)
Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenu.kt:143]: This is a great discussion and I learned something new about `Surface` as well. The simplicity of having a `Surface` providing content color is so convenient. I feel that the suggested approach works for now; we might want to discuss further once we decide to port everything to Compose (if ever).

File: docs/compose.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+Inside a specific feature, we can follow the same structure `ui/[feature]/components`, etc. 
+
+
+# Managing State 👩‍💻 <a name="managing-state"></a>
+
+Managing state properly in Compose is key to updating the UI as expected and making composable functions as reusable as possible. Some key takes to managing state: 
+
+- Recommended talk on Compose [state](https://www.youtube.com/watch?v=rmv2ug-wW4U&ab_channel=AndroidDevelopers)
+- Best practices on handling state in `@Composable` functions: 
+	- Apply state hoisting whenever possible. State hoisting is basically moving all private state out of the @Composable functions to make composable functions stateless. Ideally, delegate data manipulation to the viewModel or at least to the parent function that is calling the @Composable function.
```
Comment (hafizrahman) [inline docs/compose.md:78]: Probably good to link to https://developer.android.com/jetpack/compose/state#state-hoisting here

File: docs/compose.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+
+# Accessibility ♿️ <a name="accessibility"></a>
+
+Most of the rules that apply for Android's view system apply for Compose UI. But is worth to highlight a few concepts and tools to handle accessibility nicely. 
+
+The key idea to make Composable components accessible is [Semantics](https://developer.android.com/jetpack/compose/semantics). Accessibility services use the `Semantics tree` to provide information to the people using the services (like talkback), and the UI testing framework uses it to make assertions. Few key takes to keep your Composable UI accessible: 
+- For composables and modifiers from the Compose foundation and material library, the Semantics tree is automatically filled and generated for you
+- When adding custom low-level composables, you will have to manually provide its semantics
+- Set `contentDescriptions` for relevant icons and images. Describe the meaning of the icon, not what it is.
```
Comment (hafizrahman) [inline docs/compose.md:99]: Love the wording of "Describe the meaning of the icon" here.
